src/platform_impl/macos/view.rs

```diff
@@ -1126,3 +1147,26 @@ fn replace_event(event: &NSEvent, option_as_alt: OptionAsAlt) -> Id<NSEvent> {
         event.copy()
     }
 }
+
+// Convert given NSRange to Rust String Range data used in this crate.
+unsafe fn get_actual_range_bounds(source_str: &str, range: NSRange) -> Option<(usize, usize)> {
+    let source_ns = NSString::from_str(source_str);
+    let source_bounds = NSRange::new(0, source_ns.length());
+    let lowerbound_u16: usize = range.location;
+    let upperbound_u16: usize = range.length + lowerbound_u16;
+    // Sanity Check.
+    let mut should_return_nil = false;
+    should_return_nil |= lowerbound_u16 > source_bounds.length;
+    should_return_nil |= upperbound_u16 < lowerbound_u16;
+    should_return_nil |= upperbound_u16 > source_bounds.length;
+    if should_return_nil {
+        return None;
+    }
+    let u16_range_from_zero_a: NSRange = NSRange::new(0, lowerbound_u16);
+    let u16_range_from_zero_b: NSRange = NSRange::new(0, upperbound_u16);
+    let sub_string_a = unsafe { source_ns.substringWithRange(u16_range_from_zero_a) };
+    let sub_string_b = unsafe { source_ns.substringWithRange(u16_range_from_zero_b) };
+    let lowerbound_u8: usize = sub_string_a.to_string().len();
+    let upperbound_u8: usize = sub_string_b.to_string().len();
```

Comment:
I think we do not need to make the meaningless allocation if we do not need.

![CleanShot 2024-04-03 at 13 03 58@2x](https://github.com/rust-windowing/winit/assets/28441561/bdeceb20-087b-4aaa-a785-9aaf4a3e320e)


```suggestion
    let lowerbound_u8: usize = sub_string_a.len();
    let upperbound_u8: usize = sub_string_b.len();
```